### PR TITLE
persist: be more lenient when applying compaction diffs to state

### PR DESCRIPTION
**Pull request by danhhz**
Because of the way Spine internally optimizes only _some_ empty batches (immediately merges them in), we can end up in a situation where a compaction res applied on another copy of state, but when we replay all of the state diffs against a new Spine locally, it merges empty batches differently in-mem and we can't exactly apply the compaction diff. Example:

- compact: [1,2),[2,3) -> [1,3)
- this spine: [0,2),[2,3) (0,1 is empty)

Ideally, we'd figure out a way to avoid this, but nothing immediately comes to mind. In the meantime, force the application (otherwise the shard is stuck and we can't do anything with it) by manually splitting the empty batch back out. For the example above:

- [0,1),[1,3) (0,1 is empty)

This can only happen when the batch needing to be split is empty, so error out if it isn't because that means something unexpected is going on.

This commit also contains an (unsuccessful) attempt to get Maelstrom to tickle this bug. Seemed worth merging anyway.

Touches #15493

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist-client/src/internal/state_diff.rs [hunk 1]
```diff
@@ -629,6 +653,120 @@ fn sniff_compaction<'a, T: Timestamp + Lattice>(
...
+                last_removed_batch.desc.upper().clone(),
+                last_removed_batch.desc.since().clone(),
+            );
+            trace.push(HollowBatch {
+                desc,
+                parts: Vec::new(),
+                len: 0,
+                runs: Vec::new(),
+            });
+            metrics.state.apply_spine_fast_path_lenient_adjustment.inc();
```
Comment (pH14) [inline src/persist-client/src/internal/state_diff.rs:746]: big fan of extending the spine metaphor to including spinal adjustments 😄

File: src/persist-client/src/internal/state_diff.rs [hunk 2]
```diff
@@ -756,3 +894,141 @@ impl<'a> Iterator for ProtoStateFieldDiffsIter<'a> {
...
+            (0, 2, 0, 100),
+            &[(0, 1, 0, 0), (1, 4, 200, 0)],
+            Ok(&[(0, 2, 0, 100), (2, 4, 200, 0)]),
+        );
+
+        // Split batch at replacement lower and upper
+        testcase(
+            (2, 6, 0, 100),
+            &[(0, 3, 0, 0), (3, 5, 0, 0), (5, 8, 0, 0)],
+            Ok(&[(0, 2, 0, 0), (2, 6, 0, 100), (6, 8, 0, 0)]),
```
Comment (pH14) [inline src/persist-client/src/internal/state_diff.rs:1017]: love these tests

File: src/persist-client/src/internal/state_diff.rs [hunk 3]
```diff
@@ -629,6 +653,120 @@ fn sniff_compaction<'a, T: Timestamp + Lattice>(
...
+) -> Result<Vec<HollowBatch<T>>, String> {
+    let mut removed_batches = Vec::new();
+    trace.retain(|b| {
+        let before_output = PartialOrder::less_equal(b.desc.upper(), replacement.desc.lower());
+        let after_output = PartialOrder::less_equal(replacement.desc.upper(), b.desc.lower());
+        let overlaps_output = !(before_output || after_output);
+        if overlaps_output {
+            removed_batches.push(b.clone());
+        }
+        !overlaps_output
```
Comment (pH14) [inline src/persist-client/src/internal/state_diff.rs:693]: for some reason I found what we're trying to do here weirdly subtle. I think it's that we're trying to `retain` something that's double negated (`!overlaps` which itself is `!(before || after)`), and preserve the ones we do remove.

maybe spell out the two paths clearly? idk, I might be losing it

```rust
if overlaps_output {
  overlapping_batches.push(b.clone());
  false
} else {
  true
}
```

File: src/persist-client/src/internal/state_diff.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/internal/state_diff.rs:746]: realized when i looked at this again that this is actually a "slow" path (it reconstructs the spine), so renamed this metric

**Review by danhhz — COMMENTED** [review]
TFTR! i'm just uncomfortable enough with this that i'd lean away from putting it in a patch release (vs letting it go through the normal release qualification next week)

**Comment by danhhz** [comment]
Replaced by https://github.com/MaterializeInc/materialize/pull/15684

**Pull request closed without merging**